### PR TITLE
Make batch compute environments updatable. Fixes GH-24374

### LINKED ISSUE #24374: r/aws_batch_compute_environment: Additional attributes can be updated in-place
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

From the [recent AWS announcement](https://aws.amazon.com/about-aws/whats-new/2022/04/aws-batch-adds-improvement-updatecomputeenvironment-api/):

> AWS Batch now expands the Compute environment (CE) configuration parameters that can be updated without having to recreate compute environments. Previously customers used UpdateComputeEnvironment API to change CE state, desired CPU count, min and max vCPU count, and service role. To update additional configuration parameters, customers have to create new CEs. With this launch, customers can use the UpdateComputeEnvironment API to dynamically update CE configuration parameters such as AMI, EC2 launch templates, instance types, and more. Rather than having to manually create a new CE to use the updated configurations, Batch will use the updated CE configuration while launching new instances and terminate instances using old CE configuration once the jobs running on them are completed.

### New or Affected Resource(s)

* [`aws_batch_compute_environment`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/batch_compute_environment)

### References

[AWS Batch User Guide](https://docs.aws.amazon.com/batch/latest/userguide/updating-compute-environments.html).

### PR DESCRIPTION
**Pull request by sworisbreathing**
### Description

AWS Batch supports [updating compute environments](https://docs.aws.amazon.com/batch/latest/userguide/updating-compute-environments.html) in certain cases. This PR allows terraform to update a batch compute environment.

### Relations

Closes #24374.
Relates #25944.
Closes https://github.com/hashicorp/terraform-provider-aws/issues/15535.

### References

* https://docs.aws.amazon.com/batch/latest/userguide/updating-compute-environments.html

### Output from Acceptance Testing

I am unable to run acceptance tests myself - the only AWS accounts I have access to are locked down such that some of the basic acceptance test scaffolding (such as creating VPCs, subnets, etc) is not permitted.

### CODE REVIEW

**Comment by sworisbreathing** [comment]
@ewbankkit @bschaatsbergen one thing I've left out of this PR is the ability to [update `imageKubernetesVersion`](https://docs.aws.amazon.com/batch/latest/userguide/compute-environments-eks.html#updating-k8s-version-ce) as the provider doesn't currently expose this config option.

This PR is already marked size/xl so I'm not sure if I should add the parameter just to make updating feature-complete, or if I should raise a separate PR to expose that setting. WDYT?

File: internal/service/batch/compute_environment.go [hunk 1]
```diff
@@ -639,6 +840,67 @@ func waitComputeEnvironmentUpdated(ctx context.Context, conn *batch.Batch, name
...
 }
 
+func isFargateType(computeResourceType string) bool {
+	if computeResourceType == batch.CRTypeFargate || computeResourceType == batch.CRTypeFargateSpot {
+		return true
+	}
+	return false
+}
+
+func isUpdatableComputeEnvironment(diff *schema.ResourceDiff) bool {
```
Comment (sworisbreathing) [inline internal/service/batch/compute_environment.go:850]: per https://docs.aws.amazon.com/batch/latest/userguide/updating-compute-environments.html
> ... An infrastructure update requires that the service role is set to AWSServiceRoleForBatch (the default) and that the allocation strategy is BEST_FIT_PROGRESSIVE or SPOT_CAPACITY_OPTIMIZED. BEST_FIT isn't supported....

File: internal/service/batch/compute_environment.go [hunk 2]
```diff
@@ -433,6 +503,55 @@ func resourceComputeEnvironmentUpdate(ctx context.Context, d *schema.ResourceDat
 	return append(diags, resourceComputeEnvironmentRead(ctx, d, meta)...)
 }
 
+func ExpandComputeEnvironmentEC2ConfigurationUpdate(ec2Configuration []interface{}, defaultImageType string) []*batch.Ec2Configuration {
```
Comment (sworisbreathing) [inline internal/service/batch/compute_environment.go:506]: Note for reviewers: At first glance this code may not appear to be correct based on the AWS documentation.

I spent about 45 minutes on a chime call with an AWS support engineer yesterday running through various update scenarios with the AWS CLI. The code implemented here is based on actual observed behavior. AWS are going to review (and hopefully update) their documentation.

File: internal/service/batch/compute_environment.go [hunk 3]
```diff
@@ -433,6 +503,55 @@ func resourceComputeEnvironmentUpdate(ctx context.Context, d *schema.ResourceDat
...
+		if v, ok := m["image_type"]; ok {
+			result.ImageType = aws.String(v.(string))
+		}
+		results = append(results, result)
+	}
+
+	return results
+}
+
+func ExpandComputeEnvironmentLaunchTemplateUpdate(launchTemplate []interface{}) *batch.LaunchTemplateSpecification {
```
Comment (sworisbreathing) [inline internal/service/batch/compute_environment.go:531]: Note for reviewers: At first glance this code may not appear to be correct based on the AWS documentation.

I spent about 45 minutes on a chime call with an AWS support engineer yesterday running through various update scenarios with the AWS CLI. The code implemented here is based on actual observed behavior. AWS are going to review (and hopefully update) their documentation.

**Comment by delagoya** [comment]
Hello, will PR be accepted?

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTARGS='-run=TestAccBatchComputeEnvironment_' PKG=batch ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/batch/... -v -count 1 -parallel 2  -run=TestAccBatchComputeEnvironment_ -timeout 180m
=== RUN   TestAccBatchComputeEnvironment_basic
=== PAUSE TestAccBatchComputeEnvironment_basic
=== RUN   TestAccBatchComputeEnvironment_disappears
=== PAUSE TestAccBatchComputeEnvironment_disappears
=== RUN   TestAccBatchComputeEnvironment_nameGenerated
=== PAUSE TestAccBatchComputeEnvironment_nameGenerated
=== RUN   TestAccBatchComputeEnvironment_namePrefix
=== PAUSE TestAccBatchComputeEnvironment_namePrefix
=== RUN   TestAccBatchComputeEnvironment_createEC2
=== PAUSE TestAccBatchComputeEnvironment_createEC2
=== RUN   TestAccBatchComputeEnvironment_CreateEC2DesiredVCPUsEC2KeyPairImageID_computeResourcesTags
=== PAUSE TestAccBatchComputeEnvironment_CreateEC2DesiredVCPUsEC2KeyPairImageID_computeResourcesTags
=== RUN   TestAccBatchComputeEnvironment_createSpot
=== PAUSE TestAccBatchComputeEnvironment_createSpot
=== RUN   TestAccBatchComputeEnvironment_CreateSpotAllocationStrategy_bidPercentage
=== PAUSE TestAccBatchComputeEnvironment_CreateSpotAllocationStrategy_bidPercentage
=== RUN   TestAccBatchComputeEnvironment_createFargate
=== PAUSE TestAccBatchComputeEnvironment_createFargate
=== RUN   TestAccBatchComputeEnvironment_createFargateSpot
=== PAUSE TestAccBatchComputeEnvironment_createFargateSpot
=== RUN   TestAccBatchComputeEnvironment_updateState
=== PAUSE TestAccBatchComputeEnvironment_updateState
=== RUN   TestAccBatchComputeEnvironment_updateServiceRole
=== PAUSE TestAccBatchComputeEnvironment_updateServiceRole
=== RUN   TestAccBatchComputeEnvironment_defaultServiceRole
=== PAUSE TestAccBatchComputeEnvironment_defaultServiceRole
=== RUN   TestAccBatchComputeEnvironment_ComputeResources_minVCPUs
=== PAUSE TestAccBatchComputeEnvironment_ComputeResources_minVCPUs
=== RUN   TestAccBatchComputeEnvironment_ComputeResources_maxVCPUs
=== PAUSE TestAccBatchComputeEnvironment_ComputeResources_maxVCPUs
=== RUN   TestAccBatchComputeEnvironment_ec2Configuration
=== PAUSE TestAccBatchComputeEnvironment_ec2Configuration
=== RUN   TestAccBatchComputeEnvironment_ec2ConfigurationPlacementGroup
=== PAUSE TestAccBatchComputeEnvironment_ec2ConfigurationPlacementGroup
=== RUN   TestAccBatchComputeEnvironment_launchTemplate
=== PAUSE TestAccBatchComputeEnvironment_launchTemplate
=== RUN   TestAccBatchComputeEnvironment_updateLaunchTemplate
=== PAUSE TestAccBatchComputeEnvironment_updateLaunchTemplate
=== RUN   TestAccBatchComputeEnvironment_UpdateSecurityGroupsAndSubnets_fargate
=== PAUSE TestAccBatchComputeEnvironment_UpdateSecurityGroupsAndSubnets_fargate
=== RUN   TestAccBatchComputeEnvironment_tags
=== PAUSE TestAccBatchComputeEnvironment_tags
=== RUN   TestAccBatchComputeEnvironment_createUnmanagedWithComputeResources
=== PAUSE TestAccBatchComputeEnvironment_createUnmanagedWithComputeResources
=== RUN   TestAccBatchComputeEnvironment_updateEC2
=== PAUSE TestAccBatchComputeEnvironment_updateEC2
=== RUN   TestAccBatchComputeEnvironment_createEC2WithoutComputeResources
=== PAUSE TestAccBatchComputeEnvironment_createEC2WithoutComputeResources
=== CONT  TestAccBatchComputeEnvironment_basic
=== CONT  TestAccBatchComputeEnvironment_defaultServiceRole
--- PASS: TestAccBatchComputeEnvironment_basic (34.63s)
=== CONT  TestAccBatchComputeEnvironment_UpdateSecurityGroupsAndSubnets_fargate
--- PASS: TestAccBatchComputeEnvironment_defaultServiceRole (39.59s)
=== CONT  TestAccBatchComputeEnvironment_createEC2WithoutComputeResources
--- PASS: TestAccBatchComputeEnvironment_createEC2WithoutComputeResources (16.11s)
=== CONT  TestAccBatchComputeEnvironment_updateEC2
--- PASS: TestAccBatchComputeEnvironment_UpdateSecurityGroupsAndSubnets_fargate (74.10s)
=== CONT  TestAccBatchComputeEnvironment_createUnmanagedWithComputeResources
--- PASS: TestAccBatchComputeEnvironment_createUnmanagedWithComputeResources (3.10s)
=== CONT  TestAccBatchComputeEnvironment_tags
--- PASS: TestAccBatchComputeEnvironment_updateEC2 (98.63s)
=== CONT  TestAccBatchComputeEnvironment_ec2ConfigurationPlacementGroup
--- PASS: TestAccBatchComputeEnvironment_tags (95.15s)
=== CONT  TestAccBatchComputeEnvironment_updateLaunchTemplate
--- PASS: TestAccBatchComputeEnvironment_ec2ConfigurationPlacementGroup (51.48s)
=== CONT  TestAccBatchComputeEnvironment_launchTemplate
--- PASS: TestAccBatchComputeEnvironment_launchTemplate (49.54s)
=== CONT  TestAccBatchComputeEnvironment_CreateEC2DesiredVCPUsEC2KeyPairImageID_computeResourcesTags
--- PASS: TestAccBatchComputeEnvironment_updateLaunchTemplate (74.79s)
=== CONT  TestAccBatchComputeEnvironment_createEC2
--- PASS: TestAccBatchComputeEnvironment_createEC2 (48.41s)
=== CONT  TestAccBatchComputeEnvironment_ComputeResources_maxVCPUs
--- PASS: TestAccBatchComputeEnvironment_ComputeResources_maxVCPUs (79.91s)
=== CONT  TestAccBatchComputeEnvironment_createSpot
--- PASS: TestAccBatchComputeEnvironment_CreateEC2DesiredVCPUsEC2KeyPairImageID_computeResourcesTags (154.62s)
=== CONT  TestAccBatchComputeEnvironment_updateServiceRole
--- PASS: TestAccBatchComputeEnvironment_updateServiceRole (84.21s)
=== CONT  TestAccBatchComputeEnvironment_ec2Configuration
--- PASS: TestAccBatchComputeEnvironment_ec2Configuration (49.46s)
=== CONT  TestAccBatchComputeEnvironment_updateState
--- PASS: TestAccBatchComputeEnvironment_updateState (68.90s)
=== CONT  TestAccBatchComputeEnvironment_nameGenerated
--- PASS: TestAccBatchComputeEnvironment_nameGenerated (50.09s)
=== CONT  TestAccBatchComputeEnvironment_namePrefix
--- PASS: TestAccBatchComputeEnvironment_namePrefix (51.18s)
=== CONT  TestAccBatchComputeEnvironment_ComputeResources_minVCPUs
--- PASS: TestAccBatchComputeEnvironment_createSpot (480.85s)
=== CONT  TestAccBatchComputeEnvironment_createFargateSpot
--- PASS: TestAccBatchComputeEnvironment_createFargateSpot (48.44s)
=== CONT  TestAccBatchComputeEnvironment_createFargate
--- PASS: TestAccBatchComputeEnvironment_ComputeResources_minVCPUs (215.72s)
=== CONT  TestAccBatchComputeEnvironment_disappears
--- PASS: TestAccBatchComputeEnvironment_createFargate (48.21s)
=== CONT  TestAccBatchComputeEnvironment_CreateSpotAllocationStrategy_bidPercentage
--- PASS: TestAccBatchComputeEnvironment_disappears (49.45s)
--- PASS: TestAccBatchComputeEnvironment_CreateSpotAllocationStrategy_bidPercentage (51.44s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/batch	1044.146s
```

**Comment by ewbankkit** [comment]
@sworisbreathing Thanks for the contribution :tada: :clap:.

**Comment by sworisbreathing** [comment]
Thanks, @ewbankkit! This is going to make our BAU operations sooo much easier!

**Pull request merged**